pennylane_forest/qvm.py

```diff
@@ -165,80 +144,66 @@ def pre_measure(self):
+            wires = self.remap_wires(operation.wires)
+
+            if i > 0 and operation.name in ("QubitStateVector", "BasisState"):
+                raise DeviceError("Operation {} cannot be used after other Operations have already been applied "
+                                  "on a {} device.".format(operation.name, self.short_name))
+
+            # Prepare for parametric compilation
+            par = []
+            for param in operation.params:
+                if isinstance(param, Variable):
+                    # Using the idx for each Variable instance
+                    parameter_string = "theta" + str(param.idx)
+                    if parameter_string not in self._parameter_map:
+                        current_ref = self.prog.declare(parameter_string, "REAL")
+                        self._parameter_reference_map[parameter_string] = current_ref
+
+                    self._parameter_map[parameter_string] = [param.val]
+
+                    # Appending the parameter reference
+                    par.append(self._parameter_reference_map[parameter_string])
+                else:
+                    par.append(param)
 
-    def probabilities(self, wires):
-        """Returns the (marginal) probabilities of the quantum state.
+            self.prog += self._operation_map[operation.name](*par, *wires)
 
-        Args:
-            wires (Sequence[int]): sequence of wires to return
-                marginal probabilities for. Wires not provided
-                are traced out of the system.
+        self.apply_rotations(rotations)
 
-        Returns:
-            array: array of shape ``[2**len(wires)]`` containing
-            the probabilities of each computational basis state
-        """
-        # create an array of size [2^len(wires), 2] to store
-        # the resulting probability of each computational basis state
-        probs = np.zeros([2 ** len(wires), 2])
-        probs[:, 0] = np.arange(2 ** len(wires))
+    def generate_samples(self):
+        if "pyqvm" in self.qc.name:
+            self._samples = self.qc.run(self.prog, memory_map=self._parameter_map)
+        else:
+            # No hash provided or parametric compilation was set to False
+            # Will compile the program
+            if self.circuit_hash is None or not self.parametric_compilation:
+                compiled_program = self.qc.compile(self.prog)
 
-        # extract the measured samples
-        res = np.array([self.state[w] for w in wires]).T
-        for i in res:
-            # for each sample, calculate which
-            # computational basis state it corresponds to
-            cb = np.sum(2 ** np.arange(len(wires) - 1, -1, -1) * i)
-            # add a tally for this computational basis state
-            # to our array of basis probabilities
-            probs[cb, 1] += 1
+            # Store the compiled program with the corresponding hash
+            elif self.circuit_hash not in self._lookup_table:
```

Comment:
Maybe it might be better to give `self._lookup_table` a more descriptive name --- how about `self._compiled_program_dict`?

Reply:
Sure, sounds really good!